### ISSUE TITLE
Before HOL - Task 5: Setup Azure DevOps Agent Step:14 

### ISSUE THREAD

**Issue by Karthikkc123** [microsoft/MCW-ML-Ops#61]
1. In [Before HOL - Task 5: Setup Azure DevOps Agent Step:14](https://github.com/microsoft/MCW-ML-Ops/blob/master/Hands-on%20lab/Before%20the%20HOL%20-%20MLOps.md), while running a below command -

   **``az extension add --name azure-cli-ml``**

   Getting an error the below error -

   **``No matching extensions for 'azure-cli-ml'. Use --debug for more information.``**

   Tried by running with available version for **azure-cli-ml** i.e, 2.0.3 

   **Note**: az and azure-cli are installed with latest version.

   Please find the below screenshot for your reference:

   ![image](https://user-images.githubusercontent.com/83394010/143484784-af407271-1382-49f7-95c0-4bd57d0c8997.png)

   I'm able to add **azure-cli-ml** by running the below command:
   **az extension add -s https://azurecliext.blob.core.windows.net/release/azure_cli_ml-1.22.0.1-py3-none-any.whl -y**

2. In [Exercise-2 Task-2](https://github.com/microsoft/MCW-ML-Ops/blob/master/Hands-on%20lab/HOL%20step-by-step%20-%20MLOps.md), Pipeline Run getting failed due to "azure-cli-ml". As shown in the below screenshot:

![image](https://user-images.githubusercontent.com/83394010/143501467-5abb00ba-5e94-4efd-b2ba-4b3876df1d5e.png)

**Comment by shirolkar** [comment]
@timahenning I will look at it early this week. Thanks.

**Comment by shirolkar** [comment]
Issue is open with Azure CLI team: https://github.com/Azure/azure-cli/issues/20524

**Comment by shirolkar** [comment]
@Karthikkc123  Downgrade Azure CLI using following command on the agent: sudo apt-get install azure-cli=2.29.1-1~bionic. After that you can install the ML extension: az extension add --name azure-cli-ml

**Comment by Karthikkc123** [comment]
Thanks @shirolkar, now I'm able install azure-cli-ml extension by using the above given command to downgrade azure-cli.

But in Exercise-2 Task-2 Step-1, **still the pipeline is getting failed with Attach folder to workspace**. Even after performing the workaround note given below step-3 throwing same error.

Please find the below screenshot for your reference:

![image](https://user-images.githubusercontent.com/83394010/144486357-9f9b8252-a090-4505-861d-f568cef771cd.png)

**Comment by shirolkar** [comment]
@Karthikkc123 This looks like a different issue - an authorization failure. I was able to run the build pipeline successfully using the documented workaround. Did you create a brand new AML workspace for this lab as described in the https://github.com/microsoft/MCW-ML-Ops/blob/master/Hands-on%20lab/Before%20the%20HOL%20-%20MLOps.md? Also, check your service connection (quick-starts-sc) - maybe delete and recreate it. Thanks. Jitendra

**Comment by DawnmarieDesJardins** [comment]
@shirolkar Hi!  There are recent comments on Azure/azure-cli#20524.  Can you please check and let me know if it's a fix for this issue?  Thank you! :)

**Comment by shirolkar** [comment]
@DawnmarieDesJardins We have a workaround that works. The Azure CLI issue is still open and I do not see a new version of Azure CLI released as yet. Once the Azure CLI issue is marked fixed / new version of Azure CLI available, I can retest without the workaround. Thanks. Jitendra

**Comment by Thrillz** [comment]
> 1. az extension add -s https://azurecliext.blob.core.windows.net/release/azure_cli_ml-1.22.0.1-py3-none-any.whl -y



> @Karthikkc123 Downgrade Azure CLI using following command on the agent: sudo apt-get install azure-cli=2.29.1-1~bionic. After that you can install the ML extension: az extension add --name azure-cli-ml

For some reason i can't seem to find any version of azure-cli=2.29.1-1~bionic

**Comment by Thrillz** [comment]
![image](https://user-images.githubusercontent.com/22027172/145004468-1d6f96dc-92e5-4855-ae7f-8cc6ce3ac0b5.png)
It say version 2.19.1 but same error happens using the sudo command for version 2.29.1

**Comment by Karthikkc123** [comment]
@shirolkar, still the Pipeline getting failed to **run training pipeline** in Exercise-2 Task-2 Step-1.

Even after performing the workaround note given below step-3 throwing same error.

Please find the below screenshot for your reference:

![image](https://user-images.githubusercontent.com/83394010/145056375-eb11d3df-4ac3-446c-8e30-b5e424afc3ea.png)

**Comment by shirolkar** [comment]
@Karthikkc123 I see that the error now is at a different step so the initial issue is resolved using the workaround. For the this particular failure did you pick up the latest files from: https://github.com/solliancenet/mcw-mlops-starter-v3. If not get the update your pipelines_master.py with the latest one from the code repo https://github.com/solliancenet/mcw-mlops-starter-v3. If that does not help, I would need the view raw log file for the error to see where it failed. Thanks. Jitendra

**Comment by shirolkar** [comment]
> ![image](https://user-images.githubusercontent.com/22027172/145004468-1d6f96dc-92e5-4855-ae7f-8cc6ce3ac0b5.png) It say version 2.19.1 but same error happens using the sudo command for version 2.29.1

@Thrillz This does not look like the MCW MLOps pipeline. Are you sure you are logging the issue with the correct repo? Thanks. Jitendra

**Comment by Karthikkc123** [comment]
@shirolkar can you please let us know what permissions do we need to build this pipeline in Exercise-2 Task-2.

We are facing the Attach folder to workspace error, as I commented previously in this issue.

And we are trying with Owner permission on subscription.

Please find the below attached raw_log for your reference:
```
2021-12-07T19:52:10.6134825Z ##[section]Starting: Attach folder to workspace
2021-12-07T19:52:10.6145301Z ==============================================================================
2021-12-07T19:52:10.6145595Z Task         : Azure CLI
2021-12-07T19:52:10.6146034Z Description  : Run Azure CLI commands against an Azure subscription in a PowerShell Core/Shell script when running on Linux agent or PowerShell/PowerShell Core/Batch script when running on Windows agent.
2021-12-07T19:52:10.6146548Z Version      : 2.1.2
2021-12-07T19:52:10.6147067Z Author       : Microsoft Corporation
2021-12-07T19:52:10.6147358Z Help         : https://docs.microsoft.com/azure/devops/pipelines/tasks/deploy/azure-cli
2021-12-07T19:52:10.6147705Z ==============================================================================
2021-12-07T19:52:10.7896478Z [command]/usr/bin/az --version
2021-12-07T19:52:11.2246589Z azure-cli                         2.29.1 *
2021-12-07T19:52:11.2246973Z 
2021-12-07T19:52:11.2247458Z core                              2.29.1 *
2021-12-07T19:52:11.2248306Z telemetry                          1.0.6
2021-12-07T19:52:11.2248558Z 
2021-12-07T19:52:11.2249403Z Python location '/opt/az/bin/python3'
2021-12-07T19:52:11.2250824Z Extensions directory '/home/demouser/.azure/cliextensions'
2021-12-07T19:52:11.2251108Z 
2021-12-07T19:52:11.2251593Z Python (Linux) 3.6.10 (default, Oct 21 2021, 06:49:22) 
2021-12-07T19:52:11.2252067Z [GCC 7.5.0]
2021-12-07T19:52:11.2252272Z 
2021-12-07T19:52:11.2252670Z Legal docs and information: aka.ms/AzureCliLegal
2021-12-07T19:52:11.2252966Z 
2021-12-07T19:52:11.2253288Z 
2021-12-07T19:52:11.2256854Z WARNING: You have 2 updates available. Consider updating your CLI installation with 'az upgrade'
2021-12-07T19:52:11.2260442Z 
2021-12-07T19:52:11.2261215Z Please let us know how we are doing: https://aka.ms/azureclihats
2021-12-07T19:52:11.2264747Z and let us know if you're interested in trying out our newest features: https://aka.ms/CLIUXstudy
2021-12-07T19:52:11.2286652Z Setting AZURE_CONFIG_DIR env variable to: /home/demouser/myagent/_work/_temp/.azclitask
2021-12-07T19:52:11.2292345Z Setting active cloud to: AzureCloud
2021-12-07T19:52:11.2354752Z [command]/usr/bin/az cloud set -n AzureCloud
2021-12-07T19:52:11.5228237Z [command]/usr/bin/az login --service-principal -u *** --password=*** --tenant *** --allow-no-subscriptions
2021-12-07T19:52:12.3263053Z [
2021-12-07T19:52:12.3263510Z   {
2021-12-07T19:52:12.3263924Z     "cloudName": "AzureCloud",
2021-12-07T19:52:12.3266195Z     "homeTenantId": "***",
2021-12-07T19:52:12.3267161Z     "id": "de92e6f0-7060-40fe-931d-866e29c05d75",
2021-12-07T19:52:12.3267639Z     "isDefault": true,
2021-12-07T19:52:12.3268046Z     "managedByTenants": [],
2021-12-07T19:52:12.3268482Z     "name": "Azure HOL 3012",
2021-12-07T19:52:12.3268876Z     "state": "Enabled",
2021-12-07T19:52:12.3269742Z     "tenantId": "***",
2021-12-07T19:52:12.3270166Z     "user": {
2021-12-07T19:52:12.3270963Z       "name": "***",
2021-12-07T19:52:12.3271389Z       "type": "servicePrincipal"
2021-12-07T19:52:12.3271978Z     }
2021-12-07T19:52:12.3272303Z   }
2021-12-07T19:52:12.3272783Z ]
2021-12-07T19:52:12.3293802Z [command]/usr/bin/az account set --subscription de92e6f0-7060-40fe-931d-866e29c05d75
2021-12-07T19:52:12.6091541Z [command]/bin/bash /home/demouser/myagent/_work/_temp/azureclitaskscript1638906730778.sh
2021-12-07T19:52:14.3813505Z ERROR: ProjectSystemException:
2021-12-07T19:52:14.3814155Z 	Message: {
2021-12-07T19:52:14.3814608Z     "error_details": {
2021-12-07T19:52:14.3815092Z         "error": {
2021-12-07T19:52:14.3816156Z             "code": "AuthorizationFailed",
2021-12-07T19:52:14.3819824Z             "message": "The client '856e6354-dbe3-458f-a17c-ba18c1781688' with object id '856e6354-dbe3-458f-a17c-ba18c1781688' does not have authorization to perform action 'Microsoft.MachineLearningServices/workspaces/read' over scope '/subscriptions/de92e6f0-7060-40fe-931d-866e29c05d75/resourceGroups/Quick-Starts/providers/Microsoft.MachineLearningServices/workspaces/quick-start-ws' or the scope is invalid. If access was recently granted, please refresh your credentials."
2021-12-07T19:52:14.3821901Z         }
2021-12-07T19:52:14.3822268Z     },
2021-12-07T19:52:14.3822688Z     "status_code": 403,
2021-12-07T19:52:14.3824140Z     "url": "https://management.azure.com/subscriptions/de92e6f0-7060-40fe-931d-866e29c05d75/resourceGroups/Quick-Starts/providers/Microsoft.MachineLearningServices/workspaces/quick-start-ws?api-version=2020-01-01"
2021-12-07T19:52:14.3824898Z }
2021-12-07T19:52:14.3825257Z 	InnerException None
2021-12-07T19:52:14.3826078Z 	ErrorResponse 
2021-12-07T19:52:14.3826437Z {
2021-12-07T19:52:14.3826770Z     "error": {
2021-12-07T19:52:14.3830517Z         "message": "{\n    \"error_details\": {\n        \"error\": {\n            \"code\": \"AuthorizationFailed\",\n            \"message\": \"The client '856e6354-dbe3-458f-a17c-ba18c1781688' with object id '856e6354-dbe3-458f-a17c-ba18c1781688' does not have authorization to perform action 'Microsoft.MachineLearningServices/workspaces/read' over scope '/subscriptions/de92e6f0-7060-40fe-931d-866e29c05d75/resourceGroups/Quick-Starts/providers/Microsoft.MachineLearningServices/workspaces/quick-start-ws' or the scope is invalid. If access was recently granted, please refresh your credentials.\"\n        }\n    },\n    \"status_code\": 403,\n    \"url\": \"https://management.azure.com/subscriptions/de92e6f0-7060-40fe-931d-866e29c05d75/resourceGroups/Quick-Starts/providers/Microsoft.MachineLearningServices/workspaces/quick-start-ws?api-version=2020-01-01\"\n}"
2021-12-07T19:52:14.3832841Z     }
2021-12-07T19:52:14.3833206Z }
2021-12-07T19:52:14.5166535Z ##[error]Script failed with exit code: 1
2021-12-07T19:52:14.5195908Z [command]/usr/bin/az account clear
2021-12-07T19:52:14.7988913Z ##[section]Finishing: Attach folder to workspace

**Comment by shirolkar** [comment]
@Karthikkc123 Do you have owner permission on AML Workspace? When you setup the service connection (https://github.com/microsoft/MCW-ML-Ops/blob/master/Hands-on%20lab/HOL%20step-by-step%20-%20MLOps.md#task-4-create-new-service-connection) please select scope level to be Machine Learning Workspace and ensure that all the selected values are correct. Based on your previous comment you were able to get past folder attach step and the pipeline failed in the run training pipeline step. Can you clarify the difference when it fails at folder attach vs run training pipeline step?

**Comment by Karthikkc123** [comment]
Yes @shirolkar we have assigned owner permission to AML workspace and followed the given instructions for creating service connection. 

Still the pipeline getting failed to attach the folder to workspace. Only once the build passed till training pipeline step, after re-running the pipeline again it is getting break down at attach folder to workspace.

**Comment by shirolkar** [comment]
Hi @Karthikkc123  – this is a permissions issue and here is the key line for you to debug:

The client '856e6354-dbe3-458f-a17c-ba18c1781688' with object id '856e6354-dbe3-458f-a17c-ba18c1781688' does not have authorization to perform action 'Microsoft.MachineLearningServices/workspaces/read' over scope '/subscriptions/de92e6f0-7060-40fe-931d-866e29c05d75/resourceGroups/Quick-Starts/providers/Microsoft.MachineLearningServices/workspaces/quick-start-ws'

Can you check your app registrations in your subscription with the above ID and sure that the app has contributor permission on the AML Workspace?

Also, can you please start a new issue for the permission error. This issue is related to not being able to install azure-cli-ml extension.

Thanks.

Jitendra

**Comment by Karthikkc123** [comment]
Hi @shirolkar now we able to build the pipeline without any issues. While performing the [Exercise-4 Task 2: Monitor the Test Deployment stage](https://github.com/microsoft/MCW-ML-Ops/blob/master/Hands-on%20lab/HOL%20step-by-step%20-%20MLOps.md) faced an error in **Test Deployment** that failed to create Docker client. 

Please find the below screenshot:

![image](https://user-images.githubusercontent.com/83394010/145257067-4a93ea4f-2483-4cb6-8068-e4fdc73e50c7.png)

Resolved the issue by running below commands in build agent vm:

`sudo snap install docker`

`sudo groupadd docker`
`sudo usermod -aG docker $USER`

In [Exercise 8 (Optional): Creating and evaluating compliance classification models](https://github.com/microsoft/MCW-ML-Ops/blob/master/Hands-on%20lab/HOL%20step-by-step%20-%20MLOps.md) while running the cell in notebook to **Restore the model from model.h5 file** getting an error that 

**OSError: SavedModel file does not exist at: ./model/model.h5/{saved_model.pbtxt|saved_model.pb}**

Please find the below screenshot for your reference:

![image](https://user-images.githubusercontent.com/83394010/145257834-a013272f-5355-455e-ab92-2a68dd1b69ca.png)

**Comment by shirolkar** [comment]
@Karthikkc123  can you run the following in the notebook. 
```
import h5py
print(h5py.__version__)
```

**Comment by Karthikkc123** [comment]
@shirolkar , please find the output in the below screenshot:

![image](https://user-images.githubusercontent.com/83394010/145267562-07992ba0-7d07-4de4-89f8-d726ba5a6f14.png)

**Comment by shirolkar** [comment]
@Karthikkc123  I am not able to reproduce the restoring the saved model issue. Can you confirm that the model was downloaded in the local folder on compute instance called "model". There should be two files in that folder: model.h5 and model.onnx. Thanks. Jitendra

**Comment by shirolkar** [comment]
@Karthikkc123 Also in the future, can you create new issues. Like I mentioned the original issue is for not able to run az extension add --name azure-cli-ml. We have been using this thread for multiple different issues. Thanks. Jitendra

**Comment by shirolkar** [comment]
@DawnmarieDesJardins  The original issue in this thread is now fixed. We can close this issue. Thanks. Jitendra
